### DISCUSSION TITLE
PyScript message passing / web worker description

### DISCUSSION THREAD

**Discussion by ntoll** [pyscript/pyscript#1141] [Proposals]
# PyScript message passing / web worker description

This is a high-level description of our proposal for consolidating how different interpreters interact with the browser's main thread from web workers.

At this moment in time we are limiting ourselves to the following context:

* Two interpreters: [Pyodide](https://pyodide.org/en/stable/) and [MicroPython](https://micropython.org/).
* A single web worker.
* A high-level architectural view (specific technical details need to be explored, refined and integrated).

In writing this document we hope to share our vision and invite feedback from everyone who cares about PyScript. This is a work in progress.

## Core concepts

Before getting to the details, and so our work is accessible to those who may not yet be familiar with various aspects of web and Python based technology, we assume knowledge of the following core concepts (follow the links to find out more and fill in gaps in your knowledge):

* [Message passing](https://en.wikipedia.org/wiki/Message_passing)
* [Web workers](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API)
* [Service workers](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API)
* [XMLHttpRequest (XHR)](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest)
* [JSON](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON)
* [Events](https://developer.mozilla.org/en-US/docs/web/api/event) and 
  [Custom Events](https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent)
* [Foreign Function Interfaces (FFI)](https://en.wikipedia.org/wiki/Foreign_function_interface)

## Motivation

PyScript is a platform to easily bring Python to a browser context. 

To this end, we use [web assembly (WASM)](https://webassembly.org/) and related technologies (such as [Emscripten](https://emscripten.org/) to allow Python to run in the browser. There are several Python interpreters compiled to WASM, and each has its strengths and weaknesses.

* Pyodide is the complete CPython runtime with added features such as compiled modules like Numpy and a way to access the main-thread/JavaScript context. But, while being complete, it is also a large download and relatively (noticeably) slow to start. **Pyodide is a target interpreter for PyScript**.
* MicroPython is a lean and efficient implementation of the Python 3 programming language that includes a small subset of the Python standard library and is optimised to run on microcontrollers and in constrained environments. But, while being small and fast to start, it isn't a "complete" Python and currently lacks a mature way to interact with the browser. **MicroPython is a target interpreter for PyScript**.
* Vanilla CPython  (the reference version of Python) has recently had work done to create a WASM compilation target. However, this is not mature and lacks many of the features found in Pyodide. _CPython may, at some future date, become a target interpreter for PyScript_.

These Python interpreters currently only run in the browser's main thread. While this gives them access to the DOM and related context, it means that when Python is working, the main thread is blocked and may potentially become unresponsive. This is a bad situation since users have no way to interact with their code, except to close the blocked browser tab.

Furthermore, since the browser works in an asynchronous and event driven manner (unlike Python's explicit and synchronous way of working), the Python interpreter finds itself in a situation that makes it hard to work in a blocking manner familiar to Python programmers.

It seems we find ourselves in a paradoxical situation: the browser's main thread can't be blocked, but Python works in a blocking / synchronous manner. But there is a way out... a way to keep the main thread unblocked, Python synchronous and retain access to the DOM and JavaScript context.

## Why we need synchronous calls from Python

There are tons of features inside the CPython internals and in the broader Python ecosystem which rely on _synchronous blocking calls_ to work. This section tries to briefly explain why they are hard to do in a browser/JS world, and why they are important.

In a normal Linux/Windows/Posix environment, it happens very often that a process is blocked on certain syscalls, usually waiting for some `I/O` to happen. For example:

* If the user calls `input()`, the process is blocked until the user presses a keystroke; in this case, the blocking syscall is
  `read()` (usually a `read()` from the file descriptor `0`, i.e. `stdin`).
* If the user calls `requests.get()`, the process is blocked until the requested resource is fetched from the network; in this
  case, the blocking syscall is likely `recv()`.
* If the user calls `pdb.set_trace()`, the execution of the Python program stops until the user enters the command `next`: at 
  a low-level, the "stop" of the execution is implemented by having the whole process waiting for a keystroke, similar to what 
  described for `input()` above.
* Graphical debuggers work in a similar way: the process is stopped until it receives a specific command from a socket, and 
  the GUI sends that command when e.g. a button is pressed. In that case, the process is blocked again on a `recv()` syscall.
* `time.sleep()` is also implemented using a blocking syscall.
* There are many more examples like these in the wider Python ecosystem.

"Blocking syscall" means that the whole process is paused (and consumes 0% CPU) by the operating system, until the OS 
receives the external input and decides to resume it. At the C level, it means that when you call a certain function (e.g. 
`read()`), the execution will not proceed until the OS resumes the process.

This model of execution doesn't play well with the browser: the browser actively forbids blocking calls in the main thread, and there is no way of emulating them. But it is possible to simulate them inside web workers, which is (one of) the aspects of this proposal.

Note that we cannot use Python-level `async`/`await` to simulate blocking calls: `async`/`await` works only for Python-level functions, but we want to be able to have blocking calls also at the C level.

## Proposal

Web workers are a well established way to offload intensive computing tasks to another thread, thus unblocking the main thread and allowing the user interface to remain responsive. We propose to run our target Python interpreters in web workers.

Yet communication between the main thread and web workers must be asynchronous, which means the synchronous manner of Python code doesn't fit this model.

The following proposal assumes that we have a primitive function called `postMessageSync()` which can be called from the Python interpreter running in the web worker. Such a function sends a message to the main thread **and waits for the result**: this makes it possible to block the execution of user code in the interpreter running in the web worker without freezing the UI, and it’s the foundation on top of which we can implement more advanced features. There are at least a couple of options for how to implement such a primitive, possibly using some existing 3rd party libraries.

## Implementing `postMessageSync()` using a service worker

Our proposal is based on a "hack" that's used to turn asynchronous web-worker communication into synchronous communication. 

In brief it works like this:

1. The interpreter in a web worker emits a `postMessageSync()` message that represents some sort of interaction with the main thread.
2. The emitted message is turned into a blocking `XHR` request to a service worker. At this point the Python interpreter is blocked from continuing since it's waiting on the response to the `XHR` request.
3. The service worker asynchronously communicates with the main thread to handle the message from the interpreter, but doesn’t return a response yet. At some point, the main thread asynchronously communicates the result to the service worker, and when this result is received, the service worker returns a response to the web worker, unblocking it. 
4. If there's an event on the main thread to be handled by the Python interpreter, the process happens in reverse. A non-blocking asynchronous message representing the event is sent from the main thread to the service worker.
5. The service worker forwards the message to the web worker and, if required, creates a `Promise` in the main thread, to resolve when the computation in the Python interpreter on the web-worker generates a result.
6. The interpreter on the web worker reacts to the event-related message in the usual Pythonic blocking manner.

The architecture of this proposal is shown in the following diagram:

![cave_painting2](https://user-images.githubusercontent.com/37602/216665489-198eb78b-e8b0-4179-bf6a-7861ad020e34.png)

**It is important to note** that the XHR version of this workflow is based on a hack, but that there is potential for a solution that works in a similar manner using the new-ish [Atomics](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Atomics) browser API.

## Solutions

The Pyodide team are (as usual) ahead of the game [with this ticket](https://github.com/pyodide/pyodide/issues/1503). As Antonio pointed out on that ticket, there already [appear to be several third party solutions](https://github.com/pyodide/pyodide/issues/1503#issuecomment-1354568229) that could help out here. The problem with this specific ticket is that it is Pyodide specific.

Others have used their own home-baked solutions that work in a similar manner to our proposed solution. For example, 
Katie Bell explains her solution in [this conference presentation](https://www.youtube.com/watch?v=-SggWFS15Do).

Since PyScript should support multiple interpreters, any solution should support both Pyodide AND MicroPython, while also making it relatively easy to port this mechanism to future interpreters we may use. To this end we could use the third party solutions mentioned in the Pyodide ticket, but in an interpreter neutral manner. In addition, there are projects [such as Partytown](https://partytown.builder.io/) that attempt to provide a well documented complete solution (that we should investigate). Others, such as [NeoMJS](https://github.com/neomjs/neo) implement a complete multi-threaded (application worker) based solution for building browser based apps on all platforms (mobile, tablet, desktop).

Clearly, this is a complicated situation with solutions of varying degrees of complexity. A large part of the upcoming work will relate to researching and planning exactly what aspects of third party solutions we wish to use, and which require us to develop something bespoke for PyScript. We may even wish to reach out to these third party projects in order to benefit from their expertise and/or collaborate on a PyScript specific solution. In an ideal world, we'd do as little as possible and rely on a well maintained, supported and documented third party solution ~ but alas, we don't live in an ideal world and unpicking how these different things interact with each other is a challenge.

## Protocol schema

Depending on the nature of our solution, we may have to specify a schema for the messages to be passed between the interpreter in the web worker and the main thread. Happily, there has already been experimental work in this area:

* @pauleveritt discusses message passing with web workers in his [Pyodide components diary](https://pyodide-components.readthedocs.io/en/latest/project_init.html), and,
* @ntoll has experimented with simple message passing with MicroPython in the [PolyPlug project](https://github.com/pyscript/polyplug).

Nevertheless, projects such as Partytown and NeoMJS already appear to have such protocols in place. In this case we should ensure we're able to do everything PyScript needs.

## Calling `postMessageSync()` from Python

For this to work, a fundamental requirement is for the interpreter to be able to emit messages to send to the service worker. The simplest solution would be to do this via `stdout` since all interpreters can `print()`. However, this has the disadvantage of re-using an aspect of the interpreter that the user (in Python) might want to use... put simply, the user's code may `print()` something which isn't intended to be a message. A work-around to monkey-patch `print()` is possible but this feels messy and a bit "magic" (in a not-good way).

The alternative is to create a means for Python to call into the JavaScript context of the web worker in which it is running. Pyodide is already able to do this thanks to its [FFI related module](https://pyodide.org/en/stable/usage/api/python-api/ffi.html). The problem with this approach is that it is specific to each interpreter. However, if we can settle on a single JavaScript function implemented for all interpreters, to call to emit the message to the service worker, then each interpreter need only have to implement a means of calling just that function.

Happily, Antonio has already [a proof-of-concept of such a thing](https://github.com/antocuni/micropython/blob/0a4f904dce8ca773adb96184a1e0f480d40749c4/ports/webassembly/modjsffi.c) for MicroPython.

```c
/* Proof of concept for a module which allows to call JS functions from
   Python.

   You can test it from a web browser, or directly from node:

   $ node ./build/micropython.js

    MicroPython v1.19.1-532-g8e912a501-dirty on 2023-02-01; JS with Emscripten
    Type "help()" for more information.
    >>> import jsffi
    >>> jsffi.console_log("hello")
    hello
    0
*/

#include <emscripten.h>
#include "py/runtime.h"

/* EM_JS is Emscripten magic.
   This defines a C function with the following signature:
      void console_log(const char *str);

   However, the *body* of the function is Javascript code. Emscripten takes
   care of doing all the glueing to connect the C and JS sides.
*/
EM_JS(void, console_log, (const char* str), {
  console.log(UTF8ToString(str));
});

/* This is the micropython-side of things. When we call
   `jsffi.console_log(...)` from Python, we end up calling this function.

   Basically, the control flow is the following:

   1. the user calls jsffi.console_log("hello") from Python

   2. we enter jsffi_console_log; `obj` is the Python-level string object

   3. we use mp_get_buffer_raise to get a char* pointer to the C buffer
      containing the string

   4. we call console_log()

   5. console_log is implemented in JS, thanks to Emscripten magic. It
      receives a "pointer" to the buffer, which is just an index into the WASM
      memory

   6. Inside console_log, we call UTF8ToString(), which is an Emscripten
      helper to convert a C pointer to char into a "proper" JS string

   7. We call the JS console.log() using the string obtained in (6)
*/
STATIC mp_obj_t jsffi_console_log(mp_obj_t obj) {
    mp_buffer_info_t bufinfo;
    mp_get_buffer_raise(obj, &bufinfo, MP_BUFFER_READ);
    console_log(bufinfo.buf);
    return mp_obj_new_int(0);
}
STATIC MP_DEFINE_CONST_FUN_OBJ_1(jsffi_console_log_obj, jsffi_console_log);


STATIC const mp_rom_map_elem_t mp_jsffi_globals_table[] = {
    { MP_ROM_QSTR(MP_QSTR___name__), MP_ROM_QSTR(MP_QSTR_jsffi) },
    { MP_ROM_QSTR(MP_QSTR_console_log), MP_ROM_PTR(&jsffi_console_log_obj) },
};

STATIC MP_DEFINE_CONST_DICT(mp_jsffi_globals, mp_jsffi_globals_table);

const mp_obj_module_t mp_module_jsffi = {
    .base = { &mp_type_module },
    .globals = (mp_obj_dict_t *)&mp_jsffi_globals,
};

// Register the module to make it available in Python.
MP_REGISTER_MODULE(MP_QSTR_jsffi, mp_module_jsffi);
```

## Next steps

This is where you come in!

* We would welcome thoughtful feedback on this proposal.
* Please investigate the third-party projects that could help us with this work.
* Experimental tests are most welcome.
* Assuming, by consensus, we're able to agree a plan for a way forward we'll need to put in the required development work
  to make this happen in PyScript (itself not an insignificant effort).

Onwards. :rocket:

**Comment by hoodmane** [comment]
> The simplest solution would be to do this via `stdout` since all interpreters can `print()`.

If print would work, you could also use 
```
FS.createDevice("/dev", "myoutput", null, output_handler);
```
 and then use 
```py
myout = open("/dev/myoutput", "wb")
print("blah", file=myout)
```
this should work identically in any Python as long as it has been compiled with Emscripten.

**Reply by hoodmane** [reply]
Another question is whether you want to support data-heavy use cases: imagine updating a Canvas2D from a worker (e.g., matplotlib). I think a slight improvement on the control flow you described is as follows:

1. The worker generates a UID for the interaction
2. The worker calls `mainThreadMessagePort.postMessage({data: whatever, uid});`
3. The worker makes a synchronous XHR with the UID in it.
4. The service worker receives the request and adds it to a Map with the UID as the key.
5. The main thread asynchronously receives the message, and handles it. When it has a response, it calls `serviceWorker.postMessage({response: whatever, uid})`. 
6. The service worker receives the message and looks up the request from the UID. It serializes the response info into a `Response` object and calls `request.respondWith(response)`.
7. The worker receives the response and is unblocked

Here's my version of your diagram:

![worker-comm](https://user-images.githubusercontent.com/8739626/216805626-d9cee816-530c-4332-bbe3-53211a06881b.png)

This allows memory buffers to be moved from in Python in the worker to the main thread with one copy by slicing the data out of the wasm memory into a TypedArray and transferring the TypedArray to the main thread.

In the opposite direction, the main thread can transfer a buffer to the service worker and the service worker can copy it into a Blob and `respondWith` the Blob. I think this requires three copies: one to convert the buffer to a Blob, one [in the `respondWith` logic](https://w3c.github.io/ServiceWorker/#fetch-event-respondwith), and one to copy the data back into wasm memory so it can be seen by Python upon receipt. Note that Atomics gets these three copies down to one, which is the main efficiency win of atomics over sync xhr + service worker.

**Reply by antocuni** [reply]
I wasn't aware of the existence of `MessageChannel` and message ports, it seems interesting.
What I don't fully understand is why sending a big buffer over a message port is faster than sending it using the normal `postMessage`. I.e.:

> This allows memory buffers to be moved from in Python in the worker to the main thread with one copy by slicing the data out of the wasm memory into a TypedArray and transferring the TypedArray to the main thread.

how would this work?

---

I think that having a fast/faster way to transfer big chunks of memory from python to the main thread is a big win, although I see it more as an optimization than a high-priority feature, but this is just IMHO.

**Reply by antocuni** [reply]
ok, never mind. I didn't see that there is a very detailed discussion below :)

**Comment by pauleveritt** [comment]
The approach by @hoodmane looks like what I implemented. (Yes, I'll shut up about what I worked on, just a small note.) 

I think Hood meant web worker, not service worker. Also, I was interested in a two-way async generator (instead of postMessage) as the "message pump" but could never get my head around it.

**Reply by ntoll** [reply]
@pauleveritt can you point directly to the source code where you do this, please...? So much of the work you do is exceptionally useful, thoughtful and ahead of the curve, but I never know where to find it..! :smiley:

**Reply by pauleveritt** [reply]
Here's the [part of the writeup](https://pyodide-components.readthedocs.io/en/latest/worker_message.html).

I'll leave that up a little longer then delete it.

**Reply by ntoll** [reply]
Thank you @pauleveritt - much appreciated. :+1: 

Please don't delete it though!!! 

I think I've read this before (when you first posted it), but often because of contextual reasons, such technical work needs re-reading and re-comprehending as one's (i.e. my own) understanding and appreciation of the situation enlarges.

Remember... I'm a Winnie the Pooh (small brained) developer. :wink:

**Reply by marimeireles** [reply]
+1 in to different contexts shine different lights over the same resolution :)
I feel like I can understand things much better than I read your work weeks ago, Paul.

**Comment by ntoll** [comment]
@hoodmane aha... I like this approach a LOT since it's interpreter agnostic and anything we build on top of it benefits the whole "Python/scripting-languages in WASM" ecosystem. (Empowerment and collaboration are important virtues I hope we can embody in PyScript and the wider Python-in-the-browser ecosystem.)

Just for my own clarity of understanding, may I explain what you're suggesting back to you so I can be certain I've understood:

* We don't need a special `postMessageSync()` function to be written for each interpreter, because we could just `.write()` (and presumably also `.read()`) data to (from) a virtual device created by Emscripten (Python's `print` function [doesn't write bytes](https://docs.python.org/3/library/functions.html?highlight=print#print), but we could use something more appropriate instead).
* Creating such a device is achieved by Emscripten with: `FS.createDevice("/dev", "myoutput", null, output_handler);` where the `output_handler` is a JavaScript function to handle anything passed into `myoutput.write()` (presumably that's where the blocking `XHR` call - or equivalent - would happen).
* In standard Python we "just" need to create a simple module (off the top of my head) to handle sending messages (note the engineering use of the word "just"):
```python
"""
Off the top of Nicholas's head (and likely broken) implementation of a postMessageSync based on Hood's suggestion.
"""
import json


_myoutput = open("/dev/myoutput", "wb")  # internal reference to the virtual device created by Emscripten.


def postMessageSync(msg):
    """
    Post's a JSON encoded msg to the JavaScript context in which the interpreter is running. JS decodes the JSON
    into a JS object, and blocks at this point while it sends the message object to the main thread in some way.
    """
    json.dump(msg, _myoutput)  # blocking call.
    return json.load(_myoutput)  # read the response from the virtual device.
```

* This is where I **think** I get what you're saying, but I want to be sure I've not misunderstood... but tl;dr there are actually **TWO** messages going on here, related to each other by a JS created UUID (acting as the message ID).
  1. An asynchronous message sent to the main thread from the web worker in the usual `postMessage` sort of a way. This message includes the message's UUID.
  2. A blocking XHR call, containing just the message's UUID, is sent to the service worker to set things up so the service worker expects a response to the first message from the main thread.
* On the main thread, when the result of the message is "asynchronously" ready, it sends the result to the service worker.
* The service worker looks up the blocking `XHR` request it received from the we worker via the UUID message ID, and responds to that currently blocked request with the result received from the main thread.
* The web worker become unblocked, the result can (perhaps) be read from the `/dev/myoutput` virtual device for processing by the Python interpreter (I realise this is me being a bit hand-wavy here).

I'd like to understand the differences / benefits / costs between your method and just using a single blocking `XHR` request. As I understand it....

In your method, there is a [zero-copy transfer](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Transferable_objects) when the web-worker calls `postMessage` to the main thread. Then the content of the blocking `XHR` request need only be the message's UUID. When the result is known, there is another zero-copy transfer when the main thread calls `postMessage` to the service worker, another when we set the response to the blocked `XHR` with `respondWith`, and finally another when the response is copied into the Python context in some way (presumably, perhaps, written to the virtual device, so Python can `.read()` it?).

When it comes to the "vanilla" single `XHR` based approach we originally envisioned, the outgoing message data needs to be serialised in some way (JSON/BLOB - it's fast????) as the body of the `XHR` request from the web-worker to the service worker. Then there's a single zero-copy transfer of the message data from the service worker to the main thread via the usual `postMessage`. The response is handled in exactly the same way.

Tallying up the transfers, and unless I've missed something, we still arrive at 4 zero copy transfers in total for each approach. 

I don't yet know enough about these two methods to be able to tell if there's any noticeable performance gain from one over another. But, having two messages feels (and this is definitely an intuition) like a more complicated approach. Given that I'm a Winnie the Pooh developer (I have a very small brain), I instinctively prefer the simpler approach (vanilla `XHR`) over the two-message approach, but remain open-minded and acknowledge my ignorance. :wink: 

Of course, in the background there's atomics... which I need to become more familiar with in order to make a more informed technical decision. I also realise that atomics is not fully supported by all browsers, meaning we'd need a fallback implementation no matter what.

This leads me to wonder if the `output_handler` and `input_handler` of the Emscripten managed virtual device can't simply be shims around some existing third party module for blocking communication between web worker and main thread - leaving that magic opaque to us (see, for example, Partytown mentioned in the original post).

Finally, @hoodmane and @pauleveritt - I really love these sorts of open, friendly, collaborative and exploratory technical discussions. Let's keep this up..! :+1: My hope that is over the next fortnight (remotely), and especially when we're all together in Portugal (are you sure you can't make it Paul..? :smile: :portugal:) we can find a "landing zone" target that defines an interpreter agnostic means of access to and manipulation of the context of the main thread from a web worker. If we manage that, I think we're in a great place for building out frameworks with PyScript.

With MicroPython :microscope: :snake:  being fast-but-incomplete on mobile/tablet devices and Pyodide :weight_lifting: :snake: being the whole-standard-Python-shebang but too-big-for-mobile we give our users a compelling and flexible platform for easy Python[s]-in-the-browser on all platforms. 

Let many flowers bloom. :rocket: :rose: What fun.

**Reply by ntoll** [reply]
I just noticed my bodged Python module is broken and would never work... but I hope you see the "gist" of what I'm getting at here. ;-)

**Reply by hoodmane** [reply]
> What fun.

Agreed it's fun to discuss this stuff.

> Just for my own clarity of understanding ... create a device

Yes, I think we understand each other in the high level details. Of course there's tons of more fiddly details to discuss.

> 4 zero copy transfers in total for each approach.

This is not correct. To get to zero copy communication, we would need a shared wasm memory. Then all communication could be zero copy. In this case what I would want to do is create a C struct with the data we want to share and a small wasm module on the host that knows how to read info out of the same C struct. Then the message posted could be just a pointer and we can block with `Atomics.wait` and `Atomics.notify`. This also requires the special headers for `SharedArrayBuffer` which cause significant trouble in certain applications (they can break existing websites in hard to debug ways).

To fix terminology, I am considering things from the perspective of Python so outgoing is Python --> main thread and incoming is main thread --> Python. Supposing we are not in this shared memory case, both outgoing and incoming messages have to be copied at least once. 

If we use Atomics but no shared memory, outgoing and incoming messages can be copied exactly once.

If we use sync xhr + service worker, outgoing data can be copied just once if we slice the data out of the wasm memory into a new buffer and transfer the buffer to the main thread using `postMessage`. If we POST the data to the service worker, this requires 2 copies:
1. `let data = Module.HEAP8.subarray(ptr, ptr+size)` to get a view of the data from the wasm memory (zero copy)
2. `XHRHttpRequest.send(data)`:  [send step 4.3.1](https://xhr.spec.whatwg.org/#ref-for-bodyinit-safely-extract) is to [safely extract](https://fetch.spec.whatwg.org/#bodyinit-safely-extract) the body of the data. In our case we are safely extracting a [BufferSource](https://fetch.spec.whatwg.org/#ref-for-BufferSource%E2%91%A0) in which case it "sets source to a **copy** of the bytes held by the object. (**copy 1**)
3. On receiving the fetch event, the ServiceWorker wants to get the data from the Request body in order to pass it to the main thread. Probably we will call `.arrayBuffer()`. The [arrayBuffer](https://fetch.spec.whatwg.org/#dom-body-arraybuffer) method runs the internal procedure [consume body](https://fetch.spec.whatwg.org/#concept-body-consume-body), step 6 of which [fully reads](https://fetch.spec.whatwg.org/#body-fully-read) the body, which copies the data into the new ArrayBuffer. (**copy 2**)
4. The service worker transfers the array buffer to the main thread (zero copy).

On the way back, we need 3 copies: on the way out we were able to copy straight from wasm memory into the request body using `subarray` but on the way back we have to first copy from the response body to an `ArrayBuffer` and then from the `ArrayBuffer` into wasm memory (using `ReadableStream` instead probably provides some performance benefit here but the way I read the spec it still involves 3 copies).

Of course, just because this is what the spec says doesn't mean that the browser engine might not be able to elide some of these copies, but I think if the spec says method A involves 2 copies and method B involves 1 copy, it's safer to just use the method that the spec says uses fewer copies.

**Reply by hoodmane** [reply]
I think if you do something like drive a video from the worker, this would make a noticeable performance benefit. Before I did this sort of analysis to Pyodide's IO, people who were trying to to video processing reported bad memory usage and stalling. Without profiling anything, I analyzed all of the steps involved and minimized the allocations / copies according to the spec, and some of these changes fixed the problems. Of course, I take this approach because I am a mathematician by training and not a good statistician, so I'm a bit scared of profiling.

**Reply by hoodmane** [reply]
> some existing third party module for blocking communication between web worker and main thread

> partytown

I think that partytown is too monolithic for our needs. It is not really a library, it is only for full applications. In particular, it controls the worker life cycle. We want to control the worker life cycle ourselves. Same thing with neomjs: it's a monolithic application. We just need a system for sending synchronous messages.

To start with, we could use https://github.com/alexmojaki/sync-message or something like that. It doesn't handle binary data efficiently, it is only really designed for string data. I think the lowest level library should be based on receiving ArrayBuffers, and then there is a separate concern of how to serialize things.

Another important point is that it should be possible to initiate multiple requests before blocking for one of them. This way requests can be resolved concurrently. I think with xhr this is not hard to do.

**Reply by ntoll** [reply]
> I am a mathematician by training and not a good statistician

I'm a tuba player by training... the only thing I can do with any confidence is blow raspberries down brass tubes, the rest is just playful learning and exploration (which is a privilege). :rofl: But as Pythagoras is claimed to have said, _"there is geometry in the humming of the strings, and there is music in the spacing of the spheres"_.

> I think the lowest level library should be based on receiving ArrayBuffers, and then there is a separate concern of how to serialize things.
>
> Another important point is that it should be possible to initiate multiple requests before blocking for one of them. This way requests can be resolved concurrently. I think with xhr this is not hard to do.

My intuition (and it is _merely an intuition_, subject to actual research, refinement and revision) is that we'll end up baking our own solution[s] to this problem space, but informed by current practices in existing third party modules and (monolithic) frameworks. 

As I currently see it we have several orthogonal aspects to the problem, that ought to be cleanly separated from each other but with cleanly defined interfaces so they can interact:

* **Interpreter interactions**: Getting messages in and out of the Python context in as interpreter-neutral manner as possible (so it'll work with Pyodide, CPython, MicroPython and perhaps other non-Python-ish scripting languages like Ruby, Lua and so on [PolyScript anyone? If we don't do it, someone else will]).
* **Message schema**: in my mind this needs defining so it is expressive yet efficient enough that we can interact with and manipulate the DOM/JS context on the main thread. In my mind's eye, much of the amazing work in Pyodide could be used to inform the technical decision making around this aspect. Furthermore, because it ought to be built so it makes no difference what the underlying interpreter might be.
* **Transport layer**: to `XHR` or not to `XHR`? Atomics or not? Service workers, oh my? Etc... This needs significant research, and is where I imagine we can learn the most from existing third party modules and projects. Again, it shouldn't make any difference what the actual messages are (nor what is generating / receiving such messages), just that the transport is efficient enough that we don't (for example) get video stuttering.
* **Main thread**: when we have `<script src="pyscript.js"></script>` what is contained therein? How does is set up all of the above? What does it do in the main thread so all the other stuff happening elsewhere in the browser is able to interact with the context of the main thread? What capabilities does it have for extending PyScript (plugins etc..?)..? This aspect of PyScript feels like the chef in the kitchen, bringing together all the ingredients together via a well understood and practised recipe. Also, to keep the food-y analogy going... if folks want to season the dish to their own particular taste, we should have some means at their disposal to do so.

I'm sure I'm missing things, but just describing such aspects is a valuable way to find the context in which to most effectively engage with the PyScript puzzle.

In my mind's eye the first thing we should do is get our story straight. The outcome being that we can describe and agree upon the general terms and approach of what we're doing. Conversations such as these (in this ticket) are how we get to that point. It's worth pointing out that I'm not precious about any of the suggestions I make... I'm merely trying to push the conversation forward so we explore and uncover all the potential bits and bobs we may want to use, re-use or learn from.

The second thing, once we have our story straight, is that we figure out the implementation details. I imagine this taking the form of collaborative "spikes" that mature into properly engineered and tested code, and along the way we ensure we get user feedback from those who will have valuable perspectives (plug-in creators, interpreter maintainers, framework designers and so on).

The third and final step, is to get the code from step two into a releasable state (and all that implies - docs, tests, packaged up properly etc etc). Then I think we're in a good place.

Thoughts..?

**Reply by hoodmane** [reply]
> we have several orthogonal aspects to the problem, that ought to be cleanly separated from each other but with cleanly defined interfaces so they can interact:

I think this is a very good breakdown of the major elements.

> Message schema

My approach to this might be a bit over-complicated but I have a prototype here: https://github.com/hoodmane/synclink based on comlink. I made a Hiwire-like map where we store DOM element references and such on the main thread and refer to by id on the worker. Then then build up call graphs which we can dispatch as synchronous or asynchronous functions. 

> Transport layer: to XHR or not to XHR? Atomics or not?

Conveniently other people already did the research here for us. The answer is that we need both xhr + service workers and Atomics. 

1. xhr + service workers runs anywhere, requires https, doesn't require custom headers, doesn't work in incognito.
2. Atomics requires https + custom headers, these headers break many existing webpages. Does work in ingonito.

If we really only want to implement one, I think it should be xhr + service workers because that is less performant but runs in more places. It's also worth noting that the shape of the two transport layers is slightly different even probably at the API level (sync xhr needs a UID, Atomics needs a response buffer) so we need to prototype both up front to make sure that the API is good enough for both.

**Comment by hoodmane** [comment]
Here's a trivial example that sets up a device for writing:

#### Set up the device:

```js
FS = pyodide.FS;
let dev = FS.makedev(FS.createDevice.major++, 0);
pyodide.FS.registerDevice(dev,  {
    write(stream, buffer, offset, length, pos, canOwn) {
        console.log(buffer.subarray(offset, offset+length));
        return length;
    }
    
});
FS.mkdev('/dev/blah', dev);
```

#### Write to the device:

```py
from array import array

f = open("/dev/blah", "wb")
a = array('i', [1,2,3,4,5])
f.write(a)
f.flush()
```

**Comment by antocuni** [comment]
sorry for chiming in late in the conversation, and let me say that I really enjoyed reading.
Let me add a couple of random notes and thoughts.

@ntoll 's breakdown of the major concerns is very good, I agree with all of it.

It seems that the majority of the conversation so far has been about low level details of how to implement the worker<=>main thread communication, and performance considerations, which is good to have. However, IMHO the concrete transport mechanism that we will end up using is less important the "getting started" with web workers. I propose to start with the simplest thing which we can do (which I seem to understand being "plain XHR", please correct me if I'm wrong) and improve it later. But @hoodmane remark is also very important: we need to make sure to have an API which works for all kind of transport mechanisms that we might want to support, so it's good to have some idea about them.

Finally, my 2 cents about performance, copying and serialization protocol. My experience tells me that we tend to overestimate the cost of copying and underestimate the cost of serializing/deserializing. Unless you need to copy many megabytes of data, `memcpy` is very fast on modern machines: so yes, if you need to send a video from the worker to the main thread we probably want to avoid copying, but for the vast majority of common use cases we can probably ignore the problem for now. (This is not to say that we should make unnecessary copies of course).

---
### OT: an optimization story

Years ago I had to optimize a system in the context of high frequency betting. I had a client and a server running on the same machine, communicating with JSON over TCP sockets, sending relatively big messages.

The first optimization attempt was to use `msgpack` instead of JSON: this gave us a pretty decent speedup (I don't remember the exact numbers, but let's say 20% speedup).

The second attempt was to avoid copying the messages: I set up a complicate system based on shared memory + some locking mechanism: it resulted in another decent speedup (another ~30%) but it was not enough to justify the complexity of the locking system.

The 3rd attempt was completely different: I switched to `capnproto`, which is a "non-serialization protocol", because the in-memory layout of objects and messages is designed in such a way that you can just `memcpy` a whole object tree over the wire, and read it on the other end without any need to deserialize. I also wrote `capnpy`, which is a super-fast implementation of capnproto for CPython and PyPy, so that the "serialization" costs were very low.

The end result was that the capnproto-based system was 10x faster than the original, even if we were still copying big buffers around.

Lesson learned: `memcpy` is fast, (de)serialization is slow, JSON is very slow.

**Reply by ntoll** [reply]
The capnproto protocol looks very interesting, especially the RPC.

**Comment by ntoll** [comment]
Pinging @dpgeorge to reveal some PyScript technical context.
